### PR TITLE
[interp] Rework the allocation of offsets for variables

### PR DESCRIPTION
**Pull request by BrzVlad**
Before this change, the offset was determined by offset on the IL stack, limiting the optimizations that can be applied to these vars. Now we allocate them separately at the end of codegen.

This also optimizes the newobj opcodes.

Aside from the newobj optimization, this change should not have any meaningful performance impact.

### CODE REVIEW

**Comment by SamMonoRT** [comment]
Contributes to https://github.com/dotnet/runtime/issues/47520

File: src/mono/mono/mini/interp/transform.c [hunk 1]
```diff
@@ -8181,6 +8310,430 @@ interp_optimize_code (TransformData *td)
...
+	td->param_area_offset = final_total_locals_size;
+	for (int i = 0; i < td->locals_size; i++) {
+		// These are allocated separately at the end of the stack
+		if (td->locals [i].flags & INTERP_LOCAL_FLAG_CALL_ARGS) {
+			td->locals [i].offset += td->param_area_offset;
+			final_total_locals_size = MAX (td->locals [i].offset + td->locals [i].size, final_total_locals_size);
+		}
+	}
+	td->total_locals_size = ALIGN_TO (final_total_locals_size, MINT_STACK_SLOT_SIZE);
+}
```
Comment (lambdageek) [inline src/mono/mono/mini/interp/transform.c:8735]: offset alloc lgtm

**Review by lambdageek — APPROVED** [review]
lgtm.

think about breaking up transform.c it's getting pretty massive

**Comment by BrzVlad** [comment]
As expected, this doesn't have an invasive performance impact, with a tendency however to improve performance https://pvscmdupload.blob.core.windows.net/drewtest/report_Daily_ca%3Dx64_cb%3Drefs-heads-vlbrez-interp-offset-allocator_co%3DUbuntu1804_cr%3Ddotnetruntime_cc%3DCompilationMode%3Dtiered-LLVM%3Dfalse-MonoAOT%3Dfalse-MonoInterpreter%3Dtrue-RunKind%3Dmicro_mono_bb%3Drefs-heads-main_2021-03-11.html

**Review by vargaz — APPROVED** [review]

**Pull request merged**